### PR TITLE
Persistent volume caching for base images

### PR DESCRIPTION
**Pull request by sharifelgamal**

### CODE REVIEW

File: cmd/warmer/cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+func addKanikoOptionsFlags(cmd *cobra.Command) {
+	RootCmd.PersistentFlags().VarP(&opts.Images, "image", "i", "Image to cache. Set it repeatedly for multiple images.")
+	RootCmd.PersistentFlags().StringVarP(&opts.CacheDir, "cache-dir", "c", "/cache", "Directory of the cache.")
+}
+
+// addHiddenFlags marks certain flags as hidden from the executor help text
+func addHiddenFlags(cmd *cobra.Command) {
+}
+
+func exit(err error) {
```
Comment (dlorenc) [inline cmd/warmer/cmd/root.go]: nit: we shouldn't need this if we return an error

File: pkg/cache/cache.go [hunk 2]
```diff
@@ -18,19 +18,25 @@ package cache
...
 	"github.com/google/go-containerregistry/pkg/v1"
 	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
 // RetrieveLayer checks the specified cache for a layer with the tag :cacheKey
 func RetrieveLayer(opts *config.KanikoOptions, cacheKey string) (v1.Image, error) {
+	if local, _ := LocalDestination(opts, cacheKey); local != nil {
```
Comment (dlorenc) [inline pkg/cache/cache.go]: Maybe "source" instead of "destination"?

File: pkg/cache/warm.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	fmt.Printf("WARMER: cachedir = %s\n", cacheDir)
+	fmt.Printf("WARMER: images = %v\n", images)
+
+	for _, image := range images {
+		cacheRef, err := name.NewTag(image, name.WeakValidation)
+		if err != nil {
+			fmt.Print(err)
+			errors.Wrap(err, fmt.Sprintf("Failed to verify image name: %s", image))
+		}
+		k8sc, err := k8schain.NewNoClient()
```
Comment (dlorenc) [inline pkg/cache/warm.go]: hmm, should we use a normal keychain if we're not on k8s?

File: deploy/Dockerfile_warmer [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Builds the static Go image to execute in a Kubernetes job
+
+FROM golang:1.10
+WORKDIR /go/src/github.com/GoogleContainerTools/kaniko
+COPY . .
+RUN make
+# Get GCR credential helper
```
Comment (dlorenc) [inline deploy/Dockerfile_warmer]: Do we need the helpers if we're relying on open auth?

Comment (sharifelgamal) [reply deploy/Dockerfile_warmer]: nope. removed.

File: pkg/util/image_util.go [hunk 5]
```diff
@@ -106,7 +105,22 @@ func cachedImage(opts *config.KanikoOptions, image string) (v1.Image, error) {
 	if err != nil {
 		return nil, err
 	}
-	cacheKey := ref.Name()
-	fmt.Printf("CACHEKEY=%s", cacheKey)
-	return cache.LocalDestination(opts, cacheKey)
+
+	k8sc, err := k8schain.NewNoClient()
```
Comment (dlorenc) [inline pkg/util/image_util.go]: I think there's an easier way to do this with no auth: https://github.com/google/go-containerregistry/blob/d54baf9aa28edb9b985a6b35b57e26e3410c2443/pkg/authn/anon.go

**Pull request merged**